test/transpile/module-resolution.spec.ts

```diff
@@ -0,0 +1,165 @@
+
+    test("emits files in outDir", () => {
+        const builder = util
+            .testProject(path.join(projectPath, "tsconfig.json"))
+            .setMainFileName(path.join(projectPath, "main.ts"))
+            .setOptions({ outDir: "tstl-out" })
+            .expectToEqual(expectedResult);
+
+        // Get the output paths relative to the project path
+        const outPaths = builder.getLuaResult().transpiledFiles.map(f => path.relative(projectPath, f.outPath));
+        expect(outPaths).toHaveLength(5);
+        expect(outPaths).toContain(path.join("tstl-out", "main.lua"));
+        // Note: outputs to lua_modules
+        expect(outPaths).toContain(path.join("tstl-out", "lua_modules", "dependency1", "index.lua"));
+        expect(outPaths).toContain(path.join("tstl-out", "lua_modules", "dependency1", "otherfile.lua"));
+        expect(outPaths).toContain(path.join("tstl-out", "lua_modules", "dependency2", "index.lua"));
+        expect(outPaths).toContain(path.join("tstl-out", "lua_modules", "dependency3", "index.lua"));
+    });
+
+    test("emits bundle in outDir", () => {
+        const mainFile = path.join(projectPath, "main.ts");
+        const builder = util
+            .testProject(path.join(projectPath, "tsconfig.json"))
+            .setMainFileName(mainFile)
+            .setOptions({ outDir: "tstl-out", luaBundle: "bundle.lua", luaBundleEntry: mainFile })
+            .expectToEqual(expectedResult);
+
+        // Get the output paths relative to the project path
+        const outPaths = builder.getLuaResult().transpiledFiles.map(f => path.relative(projectPath, f.outPath));
+        expect(outPaths).toHaveLength(1);
+        expect(outPaths).toContain(path.join("tstl-out", "bundle.lua"));
+    });
+});
+
+describe("module resolution with sourceDir", () => {
+    const projectPath = path.resolve(__dirname, "module-resolution", "project-with-sourceDir");
+    const expectedResult = {
+        result: "dependency3",
+        functionInSubDir: "non-node_modules import",
+        functionReExportedFromSubDir: "nested func result",
+        nestedFunctionInSubDirOfSubDir: "nested func result",
+        nestedFunctionUsingFunctionFromParentDir: "nested func: non-node_modules import 2",
+    };
+
+    test("can resolve dependencies with sourceDir", () => {
+        util.testProject(path.join(projectPath, "tsconfig.json"))
+            .setMainFileName(path.join(projectPath, "src", "main.ts"))
+            .setOptions({ outDir: "tstl-out" })
+            .expectToEqual(expectedResult);
+    });
+
+    test("can resolve dependencies and bundle files with sourceDir", () => {
+        const mainFile = path.join(projectPath, "src", "main.ts");
+        util.testProject(path.join(projectPath, "tsconfig.json"))
+            .setMainFileName(mainFile)
+            .setOptions({ luaBundle: "bundle.lua", luaBundleEntry: mainFile })
+            .expectToEqual(expectedResult);
+    });
+});
```

Comment:
Maybe add a few tests, with filenames that have uppercase chars as well.

Reply:
the lua source test has file names with upper case names in it. I did some further tests and everything seems to work  fine, adding more tests seems useless. I even tried some cases where the module name cases mismatched the files on disk. This worked fine on windows (paths are case insensitive) and failed on linux, but on linux TS will also fail to resolve these modules, giving you an error.